diff --git a/oslo_log/log.py b/oslo_log/log.py
index 8d8c240..1b19072 100644
--- a/oslo_log/log.py
+++ b/oslo_log/log.py
@@ -285,10 +285,10 @@ def _setup_logging_from_conf(conf, project, version):
         facility = _find_facility(conf.syslog_log_facility)
         # TODO(bogdando) use the format provided by RFCSysLogHandler after
         # existing syslog format deprecation in J
-        syslog = handlers.OSSysLogHandler(
+        syslog_handler = handlers.OSSysLogHandler(
             facility=facility,
             use_syslog_rfc_format=conf.use_syslog_rfc_format)
-        log_root.addHandler(syslog)
+        log_root.addHandler(syslog_handler)

     datefmt = conf.log_date_format
     for handler in log_root.handlers:

